src/compiler/preprocess/index.ts

```diff
@@ -1,218 +1,168 @@
-
-	const name_idx = generator._names.toArray()
-		.reduce((acc, val, idx) => (acc[val] = idx, acc), {});
-
-	const mappings = generator._mappings.toArray();
-	result_line = converted_mappings[0];
-
-	for (let i = 0, len = mappings.length; i < len; i++) {
-		mapping = mappings[i];
-
-		if (mapping.generatedLine > previous_generated_line) {
-			while (mapping.generatedLine > previous_generated_line) {
-				converted_mappings.push([]);
-				previous_generated_line++;
-			}
-			result_line = converted_mappings[mapping.generatedLine - 1]; // line is one-based
-		} else if (i > 0) {
-			const previous_mapping = mappings[i - 1];
-			if (
-				// sorted by selectivity
-				mapping.generatedColumn === previous_mapping.generatedColumn &&
-				mapping.originalColumn === previous_mapping.originalColumn &&
-				mapping.name === previous_mapping.name &&
-				mapping.generatedLine === previous_mapping.generatedLine &&
-				mapping.originalLine === previous_mapping.originalLine &&
-				mapping.source === previous_mapping.source
-		) {
-				continue;
-			}
+	update_source({ string: source, map, dependencies }: SourceUpdate) {
+		this.source = source;
+		this.get_location = getLocator(source);
+
+		if (map) {
+			this.sourcemap_list.unshift(map);
 		}
-		result_line.push([mapping.generatedColumn]);
-		result_segment = result_line[result_line.length - 1];
-
-		if (mapping.source != null) {
-			result_segment.push(...[
-				source_idx[mapping.source],
-				mapping.originalLine - 1, // line is one-based
-				mapping.originalColumn
-			]);
-			if (mapping.name != null) {
-				result_segment.push(name_idx[mapping.name]);
-			}
+
+		if (dependencies) {
+			this.dependencies.push(...dependencies);
 		}
 	}
 
-	const map = {
-		version: generator._version,
-		sources: generator._sources.toArray(),
-		names: generator._names.toArray(),
-		mappings: converted_mappings
-	};
-	if (generator._file != null) {
-		(map as any).file = generator._file;
+	to_processed(): Processed {
+		// Combine all the source maps for each preprocessor function into one
+		const map: RawSourceMap = combine_sourcemaps(this.file_basename, this.sourcemap_list);
+
+		return {
+			// TODO return separated output, in future version where svelte.compile supports it:
+			// style: { code: styleCode, map: styleMap },
+			// script { code: scriptCode, map: scriptMap },
+			// markup { code: markupCode, map: markupMap },
+
+			code: this.source,
+			dependencies: [...new Set(this.dependencies)],
+			map: map as object,
+			toString: () => this.source
+		};
 	}
-	// not needed: map.sourcesContent and map.sourceRoot
-	return map;
 }
 
 /**
- * Convert a preprocessor output and its leading prefix and trailing suffix into StringWithSourceMap
+ * Convert preprocessor output for the tag content into StringWithSourceMap
  */
-function get_replacement(
-	file_basename: string,
-	offset: number,
-	get_location: ReturnType<typeof getLocator>,
-	original: string,
+function processed_content_to_sws(
```

Comment:
`sws` wasn't an obvious acronym to me - though I was able to figure it out after a little bit based on the function doc and return type of the function. Maybe we could rename `StringWithSourcemap` to something shorter like `MappedCode` so that there's not as much need to use the `sws` acronym throughout?

Reply:
`MappedCode` is much better! I'll rename it. Generally conjunctions ("with", "and") in variable and function names is a bit of a code smell.